feat: Material support

schinkowski opened the pull request.

Commit:
Adding improved material support.

Commit:
WIP material support.

Commit:
WIP material support

Commit:
More texture support

Commit:
WIP standard built-in pipeline support.

Commit:
Adding basic material support.
Adding material test scene.

Commit:
Merge branch 'dev' of https://github.com/unity3d-jp/MeshSync into material_support

# Conflicts:
#	Runtime/Plugins/x86_64/mscore.dll

SonnyCampbellUnity commented on Runtime/Shaders/ShaderHelper.cs:
```diff
@@ -0,0 +1,299 @@
+
+            const int dim = 8;
+
+            var pixels = Enumerable.Repeat(new Color(value, value, value, value), dim * dim).ToArray();
```
Should the `value` for alpha be 1 here?

schinkowski replied:
I think it doesn't make any difference either way, this generates a 1D texture so only red is used.

seandillon92 commented on Plugin~/Src/MeshSync/Include/MeshSync/SceneGraph/msTexture.h:
```diff
@@ -43,6 +43,7 @@ enum class TextureType
 {
     Default,
     NormalMap,
+    NonColor
```
NonColor is anything but Albedo, Normal or Diffuse?

schinkowski replied:
NonColor is a texture that's not set to sRGB colour space in blender.

SonnyCampbellUnity commented on Runtime/Scripts/BaseMeshSync.Textures.cs:
```diff
@@ -0,0 +1,183 @@
+        _MetallicGlossMap,
+        _BaseMap,
+        _MaskMap
+    };
```
Can we move these out into a separate class, it's kind of weird seeing `BaseMeshSync._MaskMap` etc. They should probably exist in their own MaterialConstants class or something

schinkowski replied:
I've moved them to MeshSyncConstants.

SonnyCampbellUnity commented on Runtime/Scripts/BaseMeshSync.cs:
```diff
@@ -2781,30 +2897,61 @@ private protected virtual void OnDisable()
-    const string _EmissionMap   = "_EmissionMap";
-    const string _EMISSION      = "_EMISSION";
+        internal const string _MaskMap = "_MaskMap";
+        internal const string _MASKMAP = "_MASKMAP";
```
Same with moving these out to a constants class. 

Also, what is the difference between _MASKMAP and _MaskMap?

schinkowski replied:
I've moved them to MeshSyncConstants.
The capital ones are the shader keywords.

SonnyCampbellUnity commented on Runtime/Shaders/ShaderHelper.cs:
```diff
@@ -0,0 +1,299 @@
+        //            return savedTexture;
+        //#endif
+        //            return renderTarget;
+        //        }
```
Do you want to keep this for debugging purposes?

schinkowski replied:
Deleted

SonnyCampbellUnity commented on Runtime/Shaders/ShaderHelper.cs:
```diff
@@ -0,0 +1,299 @@
+        //            return renderTarget;
+        //        }
+
+        #region Internals
```
It would make things a bit cleaner/easier to understand if seaprate the baking logic and this `Internals` region functionality into a separate classes/files

schinkowski replied:
Done

SonnyCampbellUnity commented on Runtime/Scripts/MeshSyncServer.Properties.cs:
```diff
@@ -301,6 +301,10 @@ internal override void ClearInstancePrefabs()
             needsClientSync = true;
         }
+
+        internal void RequestPythonCallback() {
```
Is this necessary?

schinkowski replied:
No, removed

SonnyCampbellUnity commented on Runtime/Scripts/BaseMeshSync.cs:
```diff
@@ -1968,7 +2084,7 @@ InstanceInfoRecord UpdateInstanceInfo(InstanceInfoData data)
             objTransform.rotation = Quaternion.LookRotation(forward, upwards);
 
-            Debug.Assert(objTransform.localToWorldMatrix == newMat, "Matrices don't match!");
+            //Debug.Assert(objTransform.localToWorldMatrix == newMat, "Matrices don't match!");
```
Is this meant to be commented out?

schinkowski replied:
No, reverted

SonnyCampbellUnity commented on Runtime/Scripts/BaseMeshSync.Textures.cs:
```diff
@@ -0,0 +1,183 @@
+                    if (tex is RenderTexture renderTarget) {
+                        var texture = new Texture2D(renderTarget.width, renderTarget.height,
+                            UnityEngine.TextureFormat.RGBA32, true);
+                        RenderTexture.active = renderTarget;
```
I think it's safer to do something like 

```
var temp = RenderTexture.active;
RenderTexture.active = renderTarget;
// ... logic
RenderTexture.active = temp;
```

schinkowski replied:
ok fixed

seandillon92 commented on Plugin~/Src/MeshSync/Utils/msMaterialExt.cpp:
```diff
@@ -33,8 +38,8 @@ void StandardMaterial::setColorMap(const TextureRecord& v)
-    if (v)
-        addProperty(MaterialProperty(_MainTex, v ));
+    // Allow null:
+    addProperty(MaterialProperty(_MainTex, v));
```
why do we need to allow null?

schinkowski replied:
To allow clearing of texture slots.

Commit:
Extracting constants

Commit:
Removing dead code.

SonnyCampbellUnity commented on Runtime/Scripts/BaseMeshSync.Textures.cs:
```diff
@@ -0,0 +1,183 @@
+
+            for (int i = pendingMaterialUpdates.Count - 1; i >= 0; i--) {
+                Tuple<Material, string> pendingMaterial = pendingMaterialUpdates[i];
+                if (SetSerializedTextureForMaterial(pendingMaterial.Item1, pendingMaterial.Item2)) {
```
I don't think I'm understanding the logic of this....

`SetSerializedTextureForMaterial` sets `EditorApplication.delayCall += UpdatePendingMaterials;`, which then removes the `UpdatePendingMaterials` callback and calls `SetSerializedTextureForMaterial` again, which seems like it could get stuck in a loop?

schinkowski replied:
It's a bit complicated. I've added some more comments.
This code runs when the meshsync server is saved. It looks at all textures on its materials that might be baked to a render texture. 
If it finds one that is a render texture, it saves that to the asset database. The asset database works asynchronously and could import the texture at some later time so the call to `LoadAssetAtPath` may return null at the time the server is saved.

To get around that, if the texture does not yet exist in the asset database, some info is added to the `pendingMaterialUpdates` list and a call to `UpdatePendingMaterials` scheduled for the next update.
When `UpdatePendingMaterials` gets called again it checks if `LoadAssetAtPath` returns a texture and if it does, it assigns the texture to the material. If it does not exist, a new call to `UpdatePendingMaterials` is scheduled.

`UpdatePendingMaterials` always unsubscribes itself at the start to avoid an infinite loop. The calls to `SetSerializedTextureForMaterial` may schedule the delayed call again. There is a safety unsubscribe there to ensure it is only scheduled once.

seandillon92 replied:
Might be better to have the caller do the registration, as suggested in another comment

schinkowski replied:
`SetSerializedTextureForMaterial` is called from multiple places, I'm trying to avoid code duplication.

SonnyCampbellUnity replied:
If you add new function to the `EditorApplication.delayCall` callbacks, during the process of the `EditorApplication.delayCall` callbacks running, does that new function get called in the current loop?

SonnyCampbellUnity replied:
i.e I'm picturing a scenario where there's a delaycall callback to `UpdatePendingMaterials`. This adds another `UpdatePendingMaterials` callback to the delaycall callbacks list, so when the first function finishes the second one runs immediately.

schinkowski replied:
No, they get called in the next delayed call.

SonnyCampbellUnity reviewed (commented):
Some questions and comments

Commit:
Refactoring

seandillon92 commented on Runtime/Scripts/BaseMeshSync.Textures.cs:
```diff
@@ -0,0 +1,183 @@
+                        texture.Apply();
+
+                        var savePath = GetSavePath(mat, textureName);
+                        TextureData.WriteToFile(savePath, texture.EncodeToPNG());
```
EncodeToPNG used to have a bug with some single channel texture formats. Lets make sure we test this.

schinkowski replied:
These will always be rgba

seandillon92 replied:
shouldn't  heightmaps be using only one channel? similarly for roughness and ao. Is this a limitation from Blender or a design choice?

schinkowski replied:
The maps that are saved here are RGBA with smoothness and metallic baked into specific channels.
There will never be height maps or any other single channel maps in these render textures.

Commit:
Reverting unwanted change

Commit:
Restore render target after changing it.

Commit:
Added comments

seandillon92 commented on Runtime/Scripts/BaseMeshSync.Textures.cs:
```diff
@@ -0,0 +1,210 @@
+            var savePath = GetSavePath(mat, textureName);
+
+            // LoadAssetAtPath can throw an exception if it's called during a domain backup but we have no way to detect that:
+            try {
+                if (AssetDatabase.LoadAssetAtPath<Texture2D>(savePath) == null) {
+                    EditorApplication.delayCall -= UpdatePendingMaterials;
+                    EditorApplication.delayCall += UpdatePendingMaterials;
+
+                    return false;
+                }
+            }
+            catch (Exception ex) {
+                return false;
+            }
```
Would it make sense if the handling was in the caller? Including both the registration to dellayCall and the exception handling

schinkowski replied:
`SetSerializedTextureForMaterial` is called from multiple places, I'm trying to avoid code duplication.

seandillon92 replied:
What about having a function that calls this one and handles exceptions or returning false. Then this function can be reused.

schinkowski replied:
What would be the value of doing that? Wouldn't that just split `SetSerializedTextureForMaterial` into 2 methods?

mloccy replied:
Would it be possible to catch the specific exception type(s) rather than Exception as this could bury other issues unintentionally, or does AssetDatabase.LoadAssetAtPath actaully throw Exception and we're just screwed?

schinkowski replied:
I've changed it to catch the UnityException instead.

seandillon92 replied:
> What would be the value of doing that? Wouldn't that just split SetSerializedTextureForMaterial into 2 methods?

Would allow us to handle the function execution as a whole. It would also break down the responsibilities.

schinkowski replied:
Not sure how useful all this is but I've extracted it to another method.

seandillon92 commented on Runtime/Scripts/BaseMeshSync.Textures.cs:
```diff
@@ -0,0 +1,210 @@
+            // Ensure the texture is up to date:
+            AssetDatabase.ImportAsset(savePath);
+
+            var savedTexture = AssetDatabase.LoadAssetAtPath<Texture2D>(savePath);
```
this call to LoadAssetAtPath could also throw an exception right?

schinkowski replied:
I've refactored it to remove the second call.

seandillon92 commented on Runtime/Scripts/BaseMeshSync.Textures.cs:
```diff
@@ -0,0 +1,210 @@
+
+                TextureImporter importer = (TextureImporter)AssetImporter.GetAtPath(savePath);
+                if (importer != null) {
+                    importer.sRGBTexture = false;
```
The editor has a config on the color space. Does this work correctly on all configs?

schinkowski replied:
The project color space doesn't have any impact on this. We just need to ensure the textures used for metallic and smoothness are in linear space. I've tested it on linear and gamma.

seandillon92 commented on Runtime/Scripts/BaseMeshSync.Textures.cs:
```diff
@@ -0,0 +1,210 @@
+        /// Tries to apply any pending material updates and unschedules itself when done.
+        /// </summary>
+        private void UpdatePendingMaterials() {
+            EditorApplication.delayCall -= UpdatePendingMaterials;
```
Do we need to unregister? from the docs, I understand that this will be called only once after being added. https://docs.unity3d.com/ScriptReference/EditorApplication-delayCall.html

schinkowski replied:
Probably not but there's no harm and it makes the code easier to read, I think.

seandillon92 replied:
A comment when you register the function might be better. If I am reading this code I would have to stop to understand why we are unregistering. We should be eliminating dead code when we can

schinkowski replied:
I've removed it now

seandillon92 commented on Runtime/Scripts/BaseMeshSync.Textures.cs:
```diff
@@ -0,0 +1,210 @@
+            bool shaderExists = shader != null;
+
+            if (shader == null) {
+#if AT_USE_HDRP
+                shader = Shader.Find("HDRP/Lit");
+#elif AT_USE_URP
+                shader = Shader.Find("Universal Render Pipeline/Lit");
+#else
+                shader = Shader.Find("Standard");
+#endif
```
Could there be a property that fetches the Lit shader name, to encapsulate this logic?

schinkowski replied:
Done

seandillon92 commented on Runtime/Scripts/BaseMeshSync.Textures.cs:
```diff
@@ -0,0 +1,210 @@
+            bool usingOverride = false;
+
+            // If the given shader name did not exist, try to set up the material to be close to the given shader name:
+            if (!shaderExists) {
+                if (shaderName.ToLower() == "glass") {
+                    mat.EnableKeyword("MESHSYNC_OVERRIDE");
+                    usingOverride = true;
+
+                    mat.EnableKeyword("_SURFACE_TYPE_TRANSPARENT");
+                    mat.SetOverrideTag("RenderType", "Transparent");
+                    mat.SetFloat("_Surface", 1);
+#if AT_USE_HDRP
+                    mat.EnableKeyword("_ENABLE_FOG_ON_TRANSPARENT");
+
+                    mat.renderQueue = 3000;
+
+                    mat.SetFloat("_SurfaceType", 1);
+                    mat.SetInt("_ZWrite", 0);
+                    mat.SetInt("_SrcBlend", (int)UnityEngine.Rendering.BlendMode.One);
+                    mat.SetInt("_DstBlend", (int)UnityEngine.Rendering.BlendMode.OneMinusSrcAlpha);
+                    mat.SetFloat("_AlphaSrcBlend", (int)UnityEngine.Rendering.BlendMode.One);
+                    mat.SetFloat("_AlphaDstBlend", (int)UnityEngine.Rendering.BlendMode.OneMinusSrcAlpha);
+                    mat.SetFloat("_ZTestDepthEqualForOpaque", 4);
+#elif AT_USE_URP
+                    mat.SetFloat("_Blend", 1);
+                    mat.DisableKeyword("_ALPHATEST_ON");
+                    mat.DisableKeyword("_ALPHABLEND_ON");
+                    mat.EnableKeyword("_ALPHAPREMULTIPLY_ON");
+                    mat.SetInt("_SrcBlend", (int)UnityEngine.Rendering.BlendMode.One);
+                    mat.SetInt("_DstBlend", (int)UnityEngine.Rendering.BlendMode.OneMinusSrcAlpha);
+#else
+                    mat.SetFloat("_Mode", 3);
+                    mat.DisableKeyword("_ALPHATEST_ON");
+                    mat.DisableKeyword("_ALPHABLEND_ON");
+                    mat.EnableKeyword("_ALPHAPREMULTIPLY_ON");
+                    mat.SetInt("_SrcBlend", (int)UnityEngine.Rendering.BlendMode.One);
+                    mat.SetInt("_DstBlend", (int)UnityEngine.Rendering.BlendMode.OneMinusSrcAlpha);
+#endif
+                }
```
This could be its own function

schinkowski replied:
Done

Commit:
Refactoring

Commit:
Enabling glass material

Commit:
Refactoring

Commit:
Refactoring.
Fix or normal map scale on HDRP.

Commit:
Catching more specific exception.

seandillon92 commented on Runtime/Scripts/BaseMeshSync.Textures.cs:
```diff
@@ -0,0 +1,218 @@
+        private List<Tuple<Material, string>> pendingMaterialUpdates = new();
+
+        /// <summary>
+        /// Returns the path where the given texture shouuld be saved.
```
typo in "should"

schinkowski replied:
Shouuld bee fiixed nowww

seandillon92 commented on Runtime/Scripts/BaseMeshSync.Textures.cs:
```diff
@@ -0,0 +1,218 @@
+        /// <param name="textureName">Name of the texture in the material</param>
+        /// <returns></returns>
+        private string GetSavePath(Material mat, string textureName) {
+            return Path.Combine(GetAssetsFolder(), $"{mat.name}_{textureName}_baked.png");
```
Does GetAssetsFolder return the path to Assets or Assets/MeshSyncAssets ?

schinkowski replied:
It returns `m_assetsFolder`, Assets/MeshSyncAssets by default

seandillon92 commented on Runtime/Scripts/BaseMeshSync.Textures.cs:
```diff
@@ -0,0 +1,218 @@
+            mat.EnableKeyword("_ALPHAPREMULTIPLY_ON");
+            mat.SetInt("_SrcBlend", (int)UnityEngine.Rendering.BlendMode.One);
+            mat.SetInt("_DstBlend", (int)UnityEngine.Rendering.BlendMode.OneMinusSrcAlpha);
+#endif
```
Could this be broken down to functions for Standard, URP and HDRP? to avoid code under preprocessor derivatives as much as possible

schinkowski replied:
Done

seandillon92 commented on Tests/Test scenes/material test scene/material_sync_test_scene_1.blend1.meta:
```diff
@@ -0,0 +1,7 @@
+  externalObjects: {}
+  userData: 
+  assetBundleName: 
+  assetBundleVariant: 
```
I don't see any blend1 files, was this meant to be checked in?

schinkowski replied:
No, deleted

Commit:
Refactoring

seandillon92 commented on Runtime/Scripts/BaseMeshSync.Textures.cs:
```diff
@@ -75,6 +78,23 @@ partial class BaseMeshSync {
         /// <param name="mat">Material to set the texture on</param>
         /// <param name="textureName">Name of the texture in the material</param>
         /// <returns></returns>
+        private bool SetSerializedTextureForMaterialOrTryAgainLater(Material mat, string textureName) {
+            if (!SetSerializedTextureForMaterial(mat, textureName)) {
+                EditorApplication.delayCall -= UpdatePendingMaterials;
+                EditorApplication.delayCall += UpdatePendingMaterials;
+
+                return false;
+            }
+
+            return true;
+        }
```
You would have to move the try catch block here. Otherwise if there is Unity Exception, it will register again, which is not the behavior in the previous version

Commit:
Refactoring

Commit:
Refactoring

seandillon92 commented on Runtime/Scripts/BaseMeshSync.Textures.cs:
```diff
@@ -0,0 +1,226 @@
+using System.Collections.Generic;
+using System.IO;
+using System.Linq;
+using UnityEditor;
```
This is in the runtime namespace, we might need to use preprocessor directives for callbacks from UnityEditor, otherwise builds will fail

schinkowski replied:
Fixed

Commit:
Tyypo fiiix.

Commit:
Deleting unused file

Commit:
Fix for runtime issues

SonnyCampbellUnity commented on Runtime/Scripts/BaseMeshSync.Textures.cs:
```diff
@@ -0,0 +1,243 @@
+                    var tex = mat.GetTexture(textureName);
+
+                    if (tex is RenderTexture renderTarget) {
+                        var texture = new Texture2D(renderTarget.width, renderTarget.height,
```
Texture2D's need to be manually destroyed with `Destroy()` after you're finished with them or they leak memory because they won't be cleaned up by the GC.

In ScanEngine we created disposable wrapper classes to handle it so that we don't forget to dispose of them https://github.cds.internal.unity3d.com/unity/ScanEngineApp/tree/ab263cbc5104dba8490f1fa6c68809b1ffa52a2c/ScanEngineApp-URP/Assets/Scripts/Common/ImageProcessing

schinkowski replied:
I've added code to release it when they're saved as texture2d in SaveMaterialRenderTexturesToAssetDatabase and no longer used.
I don't think we can wrap them in a disposable class because they are only set on the material and we don't control its lifecycle except replacing them with textures saved to the asset database. We have no control over the user replacing them with their own texture or removing them from the material.

SonnyCampbellUnity commented on Runtime/Shaders/ComputeShaderHelper.cs:
```diff
@@ -0,0 +1,55 @@
+            if (renderTarget == null ||
+                renderTarget.width != maxTextureSize.x ||
+                renderTarget.height != maxTextureSize.y) {
+                renderTarget = new RenderTexture(maxTextureSize.x, maxTextureSize.y, 32) {
```
RenderTextures need to be manually disposed of or they leak memory - see comment about Texture2Ds

There is a RenderTextureDisposeable implementation we used for ScanEngine that could be helpful

https://github.cds.internal.unity3d.com/unity/ScanEngineApp/tree/ab263cbc5104dba8490f1fa6c68809b1ffa52a2c/ScanEngineApp-URP/Assets/Scripts/Common/ImageProcessing

schinkowski replied:
I've added code to release it when they're saved as texture2d in `SaveMaterialRenderTexturesToAssetDatabase` and no longer used.
I don't think we can wrap them in a disposable class because they are only set on the material and we don't control its lifecycle except replacing them with textures saved to the asset database. We have no control over the user replacing them with their own texture or removing them from the material.

SonnyCampbellUnity commented on Runtime/Shaders/MapsBaker.cs:
```diff
@@ -0,0 +1,230 @@
+
+            var pixels = Enumerable.Repeat(new Color(value, value, value, value), dim * dim).ToArray();
+
+            texture = new Texture2D(dim, dim, UnityEngine.TextureFormat.RFloat, false, true);
```
Need to ensure this is correctly disposed of

schinkowski replied:
This is a Texture2DDisposable now

seandillon92 commented:
Should Tests/Test scenes/material test scene/material_sync_test_scene_1.blend be part of MeshSyncDCCPlugins?

Commit:
Adding texture wrapper to release resources.
Refactoring.

Commit:
Compile fixes for hdrp

Commit:
Fix for unity 2020

Commit:
Using 0.5 as default for AO and detail mask in mask map.
Fix for resetting transparency.
Updating test scene to test reroutes, add shader and mix shader.

Commit:
Removing test scene.

schinkowski commented:
> Should Tests/Test scenes/material test scene/material_sync_test_scene_1.blend be part of MeshSyncDCCPlugins?

Yeah probably, I've moved it now

Commit:
Release rendertextures after replacing them with texture2d.

Commit:
Updating comments

SonnyCampbellUnity commented on Runtime/Scripts/BaseMeshSync.Textures.cs:
```diff
@@ -0,0 +1,245 @@
+                    if (tex is RenderTexture renderTarget) {
+                        using (var texture = new Texture2DDisposable(
+                                   new Texture2D(renderTarget.width, renderTarget.height,
+                                   UnityEngine.TextureFormat.RGBA32, true))) {
```
Not a blocker, but in future probably worth adding a Texture2DDisposable constructor that takes the parameters for the Texture2D so you don't need to worry about creating a `new Texture2d()` here too

schinkowski replied:
I thought about that but there are a lot of constructors for Texture2D, I don't really want to rewrite them all to pass that down.

SonnyCampbellUnity commented on Runtime/Shaders/MapsBaker.cs:
```diff
@@ -0,0 +1,235 @@
+            destMat.SetTexture(channelName, texture);
+
+            glossTexture.Dispose();
+            rgbTexture.Dispose();
```
Do we need a null check here? Could the never get set?

schinkowski replied:
This is a Texture2DDisposable which is never null and internally releases the texture if it was set to do so.

Commit:
Fix for render textures not being released.

Commit:
Release textures that were created.

Commit:
Merge branch 'dev' of https://github.com/unity3d-jp/MeshSync into material_support

sindharta commented on Runtime/Scripts/Extensions/MaterialExtensions.cs:
```diff
@@ -0,0 +1,23 @@
+﻿using UnityEngine;
+
+namespace Unity.MeshSync {
+public static class MaterialExtensions {
```
Please make this to internal

schinkowski replied:
Done

sindharta commented on Runtime/Scripts/msCoreAPI.cs:
```diff
@@ -564,6 +570,14 @@ public struct TextureRecord {
+    public string stringValue {
+        get {
+            StringBuilder ret = new StringBuilder(arrayLength + 1);
+            msMaterialPropCopyData(self, ret);
```
Doesn't this directlly call `memcpy` on the C++ side against a StringBuilder datatype ?
Considering that the implementation of StringBuilder should be abstracted, is this really safe ?

schinkowski replied:
It never caused an issue but it's not used for now so I've commented it out.

sindharta commented on Runtime/Scripts/Texture2DDisposable.cs:
```diff
@@ -0,0 +1,47 @@
+                return;
+            }
+
+            if (Application.isPlaying) {
```
Please use `FilmInternalUtilities.ObjectUtility.Destroy();`

schinkowski replied:
Thanks, didn't know we had that.

schinkowski replied:
I've found an issue with that and had to revert it. The film utilities version uses the undo implementation, which causes an infinite loop in our case because we call this during serialization, which causes serialization to happen.

sindharta commented on Runtime/Scripts/Texture2DDisposable.cs:
```diff
@@ -0,0 +1,47 @@
+    internal class Texture2DDisposable : IDisposable {
+        public Texture2D Texture { get; private set; }
+
+        public bool DoDispose { get; }
```
public does not seem to be necessary here.

schinkowski replied:
Made it private.

sindharta commented on Runtime/Scripts/MeshSyncConstants.cs:
```diff
@@ -1,14 +1,59 @@
+        internal const string DEFAULT_ASSETS_PATH = "Assets/MeshSyncAssets";
+
+        // keyword strings
+        internal const string _Color = "_Color";
```
We should use `Shader.PropertyToID` in general rather than using string for performance reason.

schinkowski replied:
Done

sindharta commented on Runtime/Scripts/Extensions/MaterialExtensions.cs:
```diff
@@ -0,0 +1,23 @@
+        var existingTexture = mat.GetTexture(name);
+        if (existingTexture is RenderTexture existingRenderTexture &&
+            texture != existingRenderTexture) {
+            existingRenderTexture.Release();
```
Why do we need to release the RenderTexture everytime here ?
If the RenderTexture is managed by another class, then we actually should not release it here. 
I would say that either
1. this `Release()` should be called by the caller
2. the caller can pass a parameter to this function to indicate if it should/should not release the RenderTexture

schinkowski replied:
This is a helper method that ensures that render textures that are replaced with other textures are released.
The class that created the render texture does not know when another texture is set on the material to replace the texture. The caller has no reference to the render texture and should not have to care what texture was previously set.

sindharta replied:
That approach assumes that we always need to release RenderTextures, which may not be true in general case.

For example, when the materials have pre-created RenderTextures, instead of the ones created automatically by the code in this PR.

In the future, someone may use this `SetTextureSafe()` function to replace a texture of a material, and find his/her RenderTextures destroyed even though that's not intended.

schinkowski replied:
Not sure how else we can handle this. This method handles our specific scenario where any render textures that were previously set need to be released to avoid memory leaks. If someone wants to hold on to the render texture they can call the other SetTexture method.

sindharta replied:
Okay, at least we should rename the function then.

`SetTextureAndReleaseExisting` perhaps ?

schinkowski replied:
Done

sindharta commented on Runtime/Scripts/BaseMeshSync.Textures.cs:
```diff
@@ -0,0 +1,245 @@
+
+            mat.renderQueue = 3000;
+
+            mat.SetFloat("_SurfaceType", 1);
```
Please use `Shader.PropertyToID` for this set functions, and put the constants in MeshSyncConstants

schinkowski replied:
Done

sindharta commented on Runtime/Scripts/BaseMeshSync.Textures.cs:
```diff
@@ -0,0 +1,245 @@
+                            TextureData.WriteToFile(savePath, texture.Texture.EncodeToPNG());
+                        }
+
+                        // Setting the texture can fail if the asset database has not imported it yet.
```
There is [ForceUpdate](https://docs.unity3d.com/ScriptReference/ImportAssetOptions.ForceUpdate.html) option for `AssetDatabase.ImportAsset`.
Perhaps we don't need this workaround if we use this option ?

schinkowski replied:
I've added the parameter to force it synchronous which should ensure the texture exists but there is another scenario where this is called during a domain backup and the texture may not exist at that point.

sindharta commented on Runtime/Scripts/BaseMeshSync.Textures.cs:
```diff
@@ -0,0 +1,245 @@
+                    importer.sRGBTexture = false;
+                    importer.alphaIsTransparency = false;
+                    importer.ignorePngGamma = true;
+                }
```
If we decide to stay with this approach (import twice), we should call `AssetDatabase.ImportAsset();` after changing TextureImporter here.
Perhaps, `AssetDatabase.Refresh();` is also required

schinkowski replied:
Why would we need to call those? The asset importer marks the assets to be reimported if it detects it's necessary.

sindharta replied:
> if it detects it's necessary.

If the asset is changed, then likely yes. But we are modifying the property of the importer here, and I am not sure if it can always detect that the importer settings are dirty.

There's also [SaveAndReimport()](https://docs.unity3d.com/ScriptReference/AssetImporter.SaveAndReimport.html) API which seems to support the idea that the Editor can't always detect dirty settings.

schinkowski replied:
Ok I've added those.

sindharta commented on Runtime/Scripts/BaseMeshSync.Textures.cs:
```diff
@@ -0,0 +1,245 @@
+                mat.EnableKeyword(MeshSyncConstants._METALLICGLOSSMAP);
+                mat.EnableKeyword(MeshSyncConstants._METALLICSPECGLOSSMAP);
+
+                TextureImporter importer = (TextureImporter)AssetImporter.GetAtPath(savePath);
```
It's probably possible to move this code to `AssetPostprocessor.OnPreprocessTexture()` and apply the settings if the path is inside `GetAssetsFolder()` and has baked postfix.
This way, we will have the texture imported once instead twice like what it does here (once that will generate the applicable TextureImporter, and second to apply the changes here).

schinkowski replied:
Relying on a filename suffix like "baked" can cause issues if users have textures with the same names.
We'd always need to import twice. `AssetImporter.GetAtPath` returns null if the asset has not been imported.
So I don't think this can really be improved, the import time is not really the issue anyway, once the texture has been sent and imported, it's never imported again.

sindharta commented on Runtime/Scripts/BaseMeshSync.Textures.cs:
```diff
@@ -0,0 +1,245 @@
+        private static void SetupGlass_HRDP(Material mat) {
+            mat.EnableKeyword("_ENABLE_FOG_ON_TRANSPARENT");
+
+            mat.renderQueue = 3000;
```
Please use a value relative to one of the [RenderQueue enum values](https://docs.unity3d.com/ScriptReference/Rendering.RenderQueue.html)

schinkowski replied:
Done

sindharta commented on Runtime/Scripts/BaseMeshSync.Textures.cs:
```diff
@@ -0,0 +1,245 @@
+
+                    if (tex is RenderTexture renderTarget) {
+                        using (var texture = new Texture2DDisposable(
+                                   new Texture2D(renderTarget.width, renderTarget.height,
```
Will we need to handle HDR textures ?

schinkowski replied:
I don't think so, we only support the default shaders with regular Texture2Ds.

sindharta commented on Runtime/Scripts/BaseMeshSync.Textures.cs:
```diff
@@ -0,0 +1,245 @@
+                if (importer != null) {
+                    importer.sRGBTexture = false;
+                    importer.alphaIsTransparency = false;
+                    importer.ignorePngGamma = true;
```
After we have copied the data from RenderTexture, I am wondering if we can set the [isReadable](https://docs.unity3d.com/ScriptReference/Texture-isReadable.html) flag to false to optimize

schinkowski replied:
It's false by default.

sindharta commented on Runtime/Shaders/MapsBaker.cs:
```diff
@@ -0,0 +1,238 @@
+    }
+
+    static MaterialPropertyData? FindMaterialProperty(List<MaterialPropertyData> materialProperties, string name) {
+        for (int i = 0; i < materialProperties.Count; i++) {
```
Is it possible to replace `List<MaterialPropertyData>` with `Dictionary<string, MaterialPropertyData>` ?
This way, we will have faster lookup.

schinkowski replied:
Performance of this was never an issue when I profiled this code. There are around 15 properties we send so it's probably not worth creating a dictionary from it at this point.

schinkowski replied:
I've changed it to a dictionary now.

sindharta commented on Runtime/Shaders/MapsBaker.cs:
```diff
@@ -0,0 +1,238 @@
+
+        if (prop.HasValue) {
+            MaterialPropertyData.TextureRecord rec = prop.Value.textureValue;
+            disposableTexture = new Texture2DDisposable(BaseMeshSync.FindTexture(rec.id, textureHolders), false);
```
If we are not going to dispose the texture later, then shouldn't we just be using Texture2D directly ?

schinkowski replied:
The texture is only disposed if it was created on the fly, otherwise that is not necessary. Texture2DDisposable keeps track of that and disposes only when required.

sindharta replied:
Keeping this comment open because it's related to another comment below in the same file.

sindharta commented on Runtime/Shaders/MapsBaker.cs:
```diff
@@ -0,0 +1,238 @@
+
+        float value = fallbackPropertyValue;
+
+        for (int i = 0; i < materialProperties.Count; i++) {
```
converting materialProperties to a Dictionary will simplify the code here.

schinkowski replied:
Done

sindharta commented on Runtime/Shaders/MapsBaker.cs:
```diff
@@ -0,0 +1,238 @@
+
+        // If there are no textures, don't bake anything, slider values can control everything:
+        if (!texturesExist ||
+            metalTexture.Texture == null ||
```
Do we need these additional checks 
`metalTexture.Texture == null || glossTexture.Texture == null` ?

It seems that these textures will never be null because of the fallbacks, and even if they are null, `texturesExist` will be false anyway.

schinkowski replied:
Yeah, I've removed that check now.

sindharta commented on Runtime/Shaders/MapsBaker.cs:
```diff
@@ -0,0 +1,238 @@
+        var pixels = Enumerable.Repeat(new Color(value, value, value, value), dim * dim).ToArray();
+
+        disposableTexture =
+            new Texture2DDisposable(new Texture2D(dim, dim, UnityEngine.TextureFormat.RFloat, false, true));
```
If we just need a black texture or a white texture as a fallback, then we can actually use [Texture2D.blackTexture](https://docs.unity3d.com/ScriptReference/Texture2D-blackTexture.html)  or [Texture2D.whiteTexture](https://docs.unity3d.com/ScriptReference/Texture2D-whiteTexture.html) instead of having to create/manage textures ourselves.

schinkowski replied:
We need greyscale values, not just black or white. We're baking a constant from a slider to the texture.

sindharta replied:
I can only see that there are three places that call `FindTexture()`, and the value of `fallBackPropertyValue` is either 0 (black), or 1 (white).
Perhaps I am missing something here ?

schinkowski replied:
The `value` that is assigned to the pixels in `FindTexture` is taken from the fallback property if it exists. 
This way we can bake float values for metallic and smoothness.

sindharta replied:
Hm ? the `fallbackPropertyValue` is always set to 0 or 1.
Can you let me know where we set `fallbackPropertyValue` to a value other than 0 nor 1 ?

schinkowski replied:
`fallbackPropertyValue` is 0 or 1 but the value used to populate the texture can come from `fallbackMaterialProperty` if it exists.

sindharta replied:
ah, I missed the `value = fallbackMaterialProperty.floatValue;` line
Ok.

sindharta commented on Runtime/Shaders/MapsBaker.cs:
```diff
@@ -0,0 +1,238 @@
+
+            // If there are no textures, don't bake anything, slider values can control everything:
+            if (!texturesExist ||
+                glossTexture.Texture == null ||
```
Similarly, do we need these additional checks 
`glossTexture.Texture == null || rgbTexture.Texture == null` ?

schinkowski replied:
Yeah, removed now

sindharta commented on Runtime/Shaders/meshsync_channel_mapping.compute:
```diff
@@ -0,0 +1,45 @@
+void smoothness_into_alpha(uint3 id : SV_DispatchThreadID)
+{
+    int2 rgbDimensions;
+    RGB.GetDimensions(rgbDimensions.x, rgbDimensions.y);
```
I think we should simply pass the dimension into a constant buffer (if we know the size) rather than forcing the shader to get the dimension every time

schinkowski replied:
We don't know the size. This could be a texture we created from a constant value or a texture the user set. The GetDimensions call is not slow, the baking is very fast.

sindharta replied:
We know the size before executing it, right ?

Accessing a constant buffer will make it faster, although the difference may not that big.
[Reference](https://gamedev.net/forums/topic/605580-performance-comparison-hlsl-texturegetdimension/4831264/).

That being said, I don't think there is any reason not to do it since the required change is not big.

schinkowski replied:
This shader is executed once, I don't think it's worth the added complexity and chance to introduce bugs that happen when the texture size changes.

sindharta replied:
It's just setting a constant buffer at the same time as setting the texture, and a better practice in general.
I don't think it's that complex, is it ?

schinkowski replied:
So you'd say it's better practice to set the texture and the texture size and potentially not setting the texture size when it's called from another place instead of getting the texture size in the shader from the texture to ensure it's up to date?

sindharta replied:
Yes, due to the following reasons:

1. it's faster
2. it's not just anywhere we can generally call SetTexture() to a computer shader anyway, and when we do that, we generally should be certain if it's intended.
3. if we wrap the SetTexture() and SetConstantBuffer() in a function, I think the chance is very low to have this problem.

schinkowski replied:
I've changed it now.

sindharta commented on Runtime/Scripts/BaseMeshSync.cs:
```diff
@@ -460,27 +464,11 @@ internal AudioClip FindAudio(int id)
-            Material ret = new Material(shader);
-            return ret;
+        private static Material CreateDefaultMaterial(string shaderName = null) {
+            Material mat = null;
```
Let's make it clear that our intention is to create a new material here, instead of relying to `UpdateShader()` to create it

schinkowski replied:
We can't create a material without a shader or another material. How would that work if we don't know the shader?

sindharta replied:
Then let's decide what shader to use first (like what it does before this change).
If there are a lot of duplicate code with `UpdateShader()`, then maybe we can  put the duplicate code in a new function.

schinkowski replied:
I've changed it now.

sindharta commented on Runtime/Scripts/BaseMeshSync.cs:
```diff
@@ -1077,99 +1071,216 @@ private protected void UpdateMaterialAssetByDefault(MaterialData src, ModelImpor
+            return null;
+        }
+
+        private void ApplyMaterialProperty(MaterialData src, Material destMat, List<TextureHolder> textureHolders, MaterialPropertyData prop, string propName, List<MaterialPropertyData> materialProperties) {
```
we are not using the `src` parameter ?

schinkowski replied:
Not anymore, removed now.

sindharta commented on Runtime/Scripts/BaseMeshSync.cs:
```diff
@@ -1077,99 +1071,216 @@ private protected void UpdateMaterialAssetByDefault(MaterialData src, ModelImpor
+#if AT_USE_HDRP
+                destMat.SetFloat(MeshSyncConstants._AlphaCutoffEnable, hasAlpha ? 1 : 0);
+                if (hasAlpha) {
+                    destMat.renderQueue = 2475;
```
Please use a value relative to one of the [RenderQueue enum](https://docs.unity3d.com/ScriptReference/Rendering.RenderQueue.html) values

schinkowski replied:
This doesn't have an enum value.

sindharta replied:
I know, but we should put it relative to the existing enum values so that it's clear what is required.
Example: `Transparent + 75`

schinkowski replied:
Fixed

sindharta commented on Runtime/Scripts/BaseMeshSync.cs:
```diff
@@ -1077,99 +1071,216 @@ private protected void UpdateMaterialAssetByDefault(MaterialData src, ModelImpor
+                destMat.SetFloat(MeshSyncConstants._AlphaClip, hasAlpha ? 1 : 0);
+#else 
+                if (hasAlpha) {
+                    destMat.SetFloat("_Mode", 1);
```
Using `Shader.PropertyToID` constant from MeshSyncConstants is ideal here too

schinkowski replied:
Fixed

sindharta commented on Runtime/Scripts/BaseMeshSync.cs:
```diff
@@ -1077,99 +1071,216 @@ private protected void UpdateMaterialAssetByDefault(MaterialData src, ModelImpor
-                        break;
-                    default: break;
+
+                        float scale = 10;
```
Why is 10 the default value ?

schinkowski replied:
It's a small fallback value in case the node setup in blender does not specify the height scale. Normally this is not used because the value comes from the displacement node.

sindharta replied:
Can you write that in a comment there in the code ?

schinkowski replied:
Done

sindharta commented on Runtime/Scripts/BaseMeshSync.cs:
```diff
@@ -1077,99 +1071,216 @@ private protected void UpdateMaterialAssetByDefault(MaterialData src, ModelImpor
+                        var heightScaleProp = FindMaterialPropertyData(materialProperties, MeshSyncConstants._Parallax);
+                        if (heightScaleProp.HasValue) {
+                            // convert from meters to centimeters and / 2 because midpoint is half of that:
+                            scale = heightScaleProp.Value.floatValue * 100 / 2;
```
Is this conversion from meter to cm DCC-dependent ?
If so, we should probably put a flag somewhere that comes from the DCC tool.

schinkowski replied:
I've moved this to convert to Unity units in blender.

Commit:
Making class internal.
Disabling unused code.

Commit:
Refactoring

Commit:
Refactoring to use property name IDs.

Commit:
Format fix

Commit:
Extracting constants

Commit:
Force synchronous import to ensure texture exists when we need it.

Commit:
Refactoring to use constant.

Commit:
Refactoring

Commit:
Removing unnecessary code.

Commit:
Refactoring

Commit:
Removing unused parameter.

Commit:
Displacement fix

Commit:
Change to use height scale directly.

Commit:
Updating dll

Commit:
Merge branch 'dev' of https://github.com/unity3d-jp/MeshSync into material_support

sindharta commented on Runtime/Scripts/BaseMeshSync.Textures.cs:
```diff
@@ -0,0 +1,258 @@
+        }
+
+        private static void SetupGlass_URP(Material mat) {
+            mat.SetFloat(MeshSyncConstants._Blend, 1);
```
I guess this is cutout mode ? Please put a comment what this mode means.

schinkowski replied:
Added a comment

sindharta commented on Runtime/Scripts/BaseMeshSync.Textures.cs:
```diff
@@ -0,0 +1,258 @@
+        }
+
+        private static void SetupGlass_Standard(Material mat) {
+            mat.SetFloat(MeshSyncConstants._Mode, 3);
```
I guess this is transparent mode ? Please put a comment what this mode means.

schinkowski replied:
done

sindharta commented on Runtime/Scripts/msCoreAPI.cs:
```diff
@@ -517,15 +523,35 @@ public struct TextureRecord {
     public string name {
-        get { return Misc.S(msMaterialPropGetName(self)); }
+        get {
+            if (self == IntPtr.Zero) {
```
This zero check is a simple change that can be made into a different PR and merged quickly.

schinkowski replied:
Done here: https://github.com/unity3d-jp/MeshSync/pull/815/files

sindharta commented on Runtime/Scripts/msCoreAPI.cs:
```diff
@@ -558,18 +596,41 @@ public struct TextureRecord {
         }
     }
 
+    // Not used yet:
```
If we are not using it in this PR, then we should put it in another PR and remove it here.

schinkowski replied:
Deleted it now

sindharta commented on Runtime/Shaders/ComputeShaderHelper.cs:
```diff
@@ -0,0 +1,55 @@
+            if (renderTarget == null ||
+                renderTarget.width != maxTextureSize.x ||
+                renderTarget.height != maxTextureSize.y) {
+                renderTarget = new RenderTexture(maxTextureSize.x, maxTextureSize.y, 32) {
```
might want to release the existing one as well which may happen for example if `renderTarget.width != maxTextureSize.x`

schinkowski replied:
That will happen in the `SetTextureSafe` call.

sindharta replied:
Hm, that can easily introduce memory leak bugs in the future when we don't call `SetTextureSafe` anymore for any reason.
Is it possible to put code to clear it here as well ?

schinkowski replied:
Done now

sindharta commented on Runtime/Shaders/MapsBaker.cs:
```diff
@@ -0,0 +1,217 @@
+                loadedShaders[name] = shaderHelper;
+            }
+#else
+// TODO: Load shader at runtime here
```
Can you convert this TODO into something like this, so that it's clear who to ask about this and when this todo was added ?
`[TODO-<person>:<date>] What to do`

schinkowski replied:
I've added a ticket number. We don't know who will do it yet.

sindharta commented on Runtime/Scripts/BaseMeshSync.cs:
```diff
@@ -1077,98 +1071,205 @@ private protected void UpdateMaterialAssetByDefault(MaterialData src, ModelImpor
+                destMat.SetFloat(MeshSyncConstants._AlphaCutoffEnable, hasAlpha ? 1 : 0);
+                if (hasAlpha) {
+                    destMat.renderQueue = 2475;
+                    destMat.SetFloat("_ZTestGBuffer", 3);
```
Please use MeshSyncConstants here too.

schinkowski replied:
Done

sindharta commented on Runtime/Scripts/BaseMeshSync.cs:
```diff
@@ -1077,98 +1071,205 @@ private protected void UpdateMaterialAssetByDefault(MaterialData src, ModelImpor
+                        destMat.SetFloat(MeshSyncConstants._HeightMax, scale);
+
+                        float amplitude = scale * 2;
+                        destMat.SetFloat(MeshSyncConstants._HeightAmplitude, amplitude * 0.01f);
```
If this code is new, can you write in a comment why we are multiplying with 2, and then with 0.01f ?
For example, if that's a unit conversion ?

schinkowski replied:
Added comments

sindharta commented on Runtime/Scripts/BaseMeshSync.Textures.cs:
```diff
@@ -0,0 +1,258 @@
+
+            Assert.IsNotNull(shader);
+
+            if (mat == null) {
```
Related to my comment in `CreateDefaultMaterial()` function.
I guess what I don't like about this function, is that it allows a new material creation instead of just updating it, which feels "hacky" to me.

schinkowski replied:
Creating a new material or updating an existing one use the same code path in `UpdateShader`. The shader can change for a material after it was created or it can be created with a specific shader and change later, this is all handled in a single method.

sindharta replied:
Right, I think this function itself is fine, but I'd prefer an explicit Material creation in `CreateDefaultMaterial()` .

Perhaps that will be possible if we split this function into two ?
1. a function that decides the shader name
2. a function that update the material properties based on the shader name

schinkowski replied:
Changed it now.

sindharta commented on Runtime/Scripts/BaseMeshSync.Textures.cs:
```diff
@@ -0,0 +1,258 @@
+            }
+
+            // If the material was set up by meshsync but doesn't need that anymore, revert it to a standard material:
+            if (!usingOverride && Array.IndexOf(mat.shaderKeywords, MeshSyncConstants.MESHSYNC_OVERRIDE) >= 0) {
```
And `CreateDefaultMaterial()` also doesn't need to call this particular code, does it ?

schinkowski replied:
Not if it uses the `UpdateShader` method to handle creation and update in a single place.

sindharta replied:
I saw your changes in `CreateDefaultMaterial()`.
Resolving this conversation.

Commit:
Ensure assets are updated after changing settings

Commit:
Ensure importer is updated.

Commit:
Adding ticket number

Commit:
Using constant

Commit:
Adding comment

Commit:
Adding comment

Commit:
Removing code we don't use yet.

Commit:
Extracting constant.

Commit:
Adding comments

Commit:
Releasing render texture

SonnyCampbellUnity commented on Runtime/Scripts/BaseMeshSync.cs:
```diff
@@ -1077,98 +1071,210 @@ private protected void UpdateMaterialAssetByDefault(MaterialData src, ModelImpor
+#if AT_USE_HDRP
+                destMat.SetFloat(MeshSyncConstants._AlphaCutoffEnable, hasAlpha ? 1 : 0);
+                if (hasAlpha) {
+                    destMat.renderQueue = (int)RenderQueue.AlphaTest + 25;
```
What is this doing, and what does `+ 25` mean? Can you use a named variable for the 25 and add a comment here?

schinkowski replied:
This is the render order that the hdrp pipeline uses when the material is set to transparent. I'm just replicating what that does. I think it would make the code even more complicated if we make a constant of every value we set on these shaders that's used in a single place.

sindharta replied:
In this case, I think using literal number values is fine as long as it's clear in which render queue the process should be executed, which in this example is `AlphaTest`.

sindharta replied:
Unless we have multiple rendering processes in the same `AlphaTest` category. I guess we don't, do we ?

schinkowski replied:
We don't have any need to set render orders for anything else. 2475 is what HDRP uses by default when the material is set to transparent, I wanted to make ours behave the same way.

Commit:
Refactoring

seandillon92 commented on Runtime/Scripts/Texture2DDisposable.cs:
```diff
@@ -0,0 +1,47 @@
+    internal class Texture2DDisposable : IDisposable {
+        public Texture2D Texture { get; private set; }
+
+        private bool DoDispose { get; }
```
is this actually used anywhere or is it dead code?

schinkowski replied:
It's used in Dispose() to check whether the texture should be disposed or not.

seandillon92 commented on Runtime/Scripts/Texture2DDisposable.cs:
```diff
@@ -0,0 +1,47 @@
+            Texture = texture;
+        }
+
+        public void ReadFromRenderTexture(RenderTexture renderTexture) {
+            var activeRenderTarget = RenderTexture.active;
+
+            RenderTexture.active = renderTexture;
+            Texture.ReadPixels(new Rect(0, 0, renderTexture.width, renderTexture.height), 0, 0);
+            Texture.Apply();
+
+            RenderTexture.active = activeRenderTarget;
+        }
```
Is this code used anywhere? since the Texture getter is public, you could have an extension on Texture2D to copy from RenderTexture

schinkowski replied:
This is used by `SaveMaterialRenderTexturesToAssetDatabase`. I don't think it's good to make it an extension method because then the user needs to reference the Texture of the object instead of keeping the logic inside this class.

seandillon92 commented on Runtime/Shaders/MapsBaker.cs:
```diff
@@ -0,0 +1,217 @@
+        }
+        else {
+            Debug.LogError(
+                $"Unknown smoothness channel, cannot bake to option: {smoothnessChannel} set in the smoothness texture channel.");
```
Most logs we have so far start with "[MeshSync]", so the user is aware where the error is coming from. Perhaps we could do the same here?

schinkowski replied:
Done

Commit:
Log update

Commit:
Adding comment

Commit:
Renaming method

Commit:
Merge branch 'dev' of https://github.com/unity3d-jp/MeshSync into material_support

# Conflicts:
#	Runtime/Plugins/x86_64/mscore.dll
#	Runtime/Scripts/msCoreAPI.cs

Commit:
Refactoring

Commit:
Docs update

Commit:
Reverting unwanted change

sindharta commented on Runtime/Scripts/BaseMeshSync.cs:
```diff
@@ -876,15 +864,32 @@ internal void AfterUpdateScene()
+
+                    // Make sure the full texture is used unless the user changed this setting before:
+                    assetExisted = true;
+                    if (!assetExisted) {
```
This seems like a new change.
`assetExisted` is always true here, isn't it ?
Can we remove the branching ?

schinkowski replied:
`assetExisted` would be false if the file was just saved and `AssetDatabase.ImportAsset` has not been called yet.

sindharta replied:
Hmm ? `assetExisted` is set to true just before the branching, which means that the branching isn't really being used, isn't it ?

schinkowski replied:
I've removed that now.

Commit:
Removing debug code.
